lib/3scale/backend/transactor/status.rb

```diff
@@ -211,11 +194,35 @@ def with_report_and_hierarchy(reports)
           end
         end
 
-        def add_plan(xml, tag, plan_name)
+        def add_plan_section(xml, tag, plan_name)
           xml << "<#{tag}>"
           xml << plan_name.to_s.encode(xml: :text) << "</#{tag}>"
         end
 
+        def add_authorize_section(xml)
+          if authorized?
+            xml << '<authorized>true</authorized>'.freeze
+          else
+            xml << '<authorized>false</authorized><reason>'.freeze
+            xml << rejection_reason_text
+            xml << '</reason>'.freeze
+          end
+        end
+
+        def add_application_section(xml)
+          redirect_uri = application.redirect_url
+          xml << "<application><id>".freeze << "#{application.id}" << "</id><key>".freeze
+          xml << "#{application.keys.first}" << "</key><".freeze
+          xml << "#{@redirect_uri_field}>#{redirect_uri}</#{@redirect_uri_field}"
+          xml << '></application>'.freeze
```

Comment:
If I do that then this literals will not be frozen even if I put a .freeze, which we were trying to do for the literal parts as mikz commented.

Reply:
@miguelsorianod but there is no good reason to have extra frozen string when you are allocating it moments before anyway

```ruby
RubyVM::InstructionSequence.compile(<<'XML').disasm
xml << "#{@redirect_uri_field}>#{redirect_uri}</#{@redirect_uri_field}";
xml << '></application>'.freeze
XML
```

```
== disasm: #<ISeq:<compiled>@<compiled>>================================
0000 trace            1                                               (   1)
0002 putself
0003 opt_send_without_block <callinfo!mid:xml, argc:0, FCALL|VCALL|ARGS_SIMPLE>, <callcache>
0006 getinstancevariable :@redirect_uri_field, <is:0>
0009 tostring
0010 putobject        ">"
0012 putself
0013 opt_send_without_block <callinfo!mid:redirect_uri, argc:0, FCALL|VCALL|ARGS_SIMPLE>, <callcache>
0016 tostring
0017 putobject        "</"
0019 getinstancevariable :@redirect_uri_field, <is:0>
0022 tostring
0023 concatstrings    5
0025 opt_ltlt         <callinfo!mid:<<, argc:1, ARGS_SIMPLE>, <callcache>
0028 pop
0029 trace            1                                               (   2)
0031 putself
0032 opt_send_without_block <callinfo!mid:xml, argc:0, FCALL|VCALL|ARGS_SIMPLE>, <callcache>
0035 opt_str_freeze   "></application>"
0037 opt_ltlt         <callinfo!mid:<<, argc:1, ARGS_SIMPLE>, <callcache>
0040 leave
```

versus 

```ruby
RubyVM::InstructionSequence.compile(<<'XML').disasm
xml << "#{@redirect_uri_field}>#{redirect_uri}</#{@redirect_uri_field}></application>"
XML
```

```
== disasm: #<ISeq:<compiled>@<compiled>>================================
0000 trace            1                                               (   1)
0002 putself
0003 opt_send_without_block <callinfo!mid:xml, argc:0, FCALL|VCALL|ARGS_SIMPLE>, <callcache>
0006 getinstancevariable :@redirect_uri_field, <is:0>
0009 tostring
0010 putobject        ">"
0012 putself
0013 opt_send_without_block <callinfo!mid:redirect_uri, argc:0, FCALL|VCALL|ARGS_SIMPLE>, <callcache>
0016 tostring
0017 putobject        "</"
0019 getinstancevariable :@redirect_uri_field, <is:0>
0022 tostring
0023 putobject        "></application>"
0025 concatstrings    6
0027 opt_ltlt         <callinfo!mid:<<, argc:1, ARGS_SIMPLE>, <callcache>
0030 leave
```

Reply:
in other words: there is one allocation already going on there with the interpolation, and freezing and adding the frozen strings adds two extra operations, whereas those strings could already be part of the allocated string in the first place.

Reply:
I understand that an allocation is less costly that an allocation + a concatenation in that specific case. My doubt was about what happened if that same literal would appear in other places of the code.
The solution then would be to freeze the '></application>' part in the other part of the code. It would not make sense to freeze it in this specific section of the code because anyway you will be doing an allocation for the redirect part. I'll change it now, thank you!

Reply:
The thing is what are you willing to optimize for. If you have the same literals sprinkled elsewhere you can make all of them frozen, only some, or none. With strings as interpreted in MRI, using `.freeze` makes the effect of interning them, which is to say they become similar to static, immutable constants. In that case, when the interpreter finds references to the same string (and also with `.freeze`*), it will refer to the interned one to save on unneeded allocations and extra copies.

But of course, having a few copies of the same string (or more accurately, substring) to save in execution speed is usually a good tradeoff. So, in some cases you can intern them, but it only makes sense if you actually save the allocation, and in that case, it's worse if it also adds an extra cost in execution time.

`*`: Ruby 2.3+ have flags for making the `String#freeze` behaviour the default with literals.

Reply:
This is 93% the same comment as above! I want my 10s back!!